lib/experimental/html/class-wp-html-walker.php

```diff
@@ -0,0 +1,1282 @@
+				$updated_attribute
+			);
+		} else {
+			/*
+			 * Create a new attribute at the tag's name end.
+			 *
+			 * Example – add attribute id="new" to <div />:
+			 *    <div/>
+			 *        ^
+			 *        start and end
+			 *    replacement: ` id="new"`
+			 *
+			 *    Result: <div id="new"/>
+			 */
+			$this->attribute_updates[ $name ] = new WP_HTML_Text_Replacement(
+				$this->tag_name_ends_at,
+				$this->tag_name_ends_at,
+				' ' . $updated_attribute
+			);
+		}
+	}
+
+	/**
+	 * Removes an attribute of the currently matched tag.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string $name The attribute name to remove.
+	 *
+	 * @throws WP_HTML_Walker_Exception Once this object was already stringified and closed.
+	 */
+	public function remove_attribute( $name ) {
+		$this->assert_not_closed();
+
+		if ( ! isset( $this->attributes[ $name ] ) ) {
+			return;
+		}
+
+		/*
+		 * Removes an existing tag attribute.
+		 *
+		 * Example – remove the attribute id from <div id="main"/>:
+		 *    <div id="initial_id"/>
+		 *         ^-------------^
+		 *         start         end
+		 *    replacement: ``
+		 *
+		 *    Result: <div />
+		 */
+		$this->attribute_updates[ $name ] = new WP_HTML_Text_Replacement(
+			$this->attributes[ $name ]->start,
+			$this->attributes[ $name ]->end,
+			''
+		);
+	}
+
+	/**
+	 * Return true when the HTML Walker is closed for further lookups and modifications.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return boolean True if the HTML Walker is closed, false otherwise.
+	 */
+	public function is_closed() {
+		return $this->closed;
+	}
+
+	/**
+	 * Adds a new class name to the currently matched tag.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string $class_name The class name to add.
+	 *
+	 * @throws WP_HTML_Walker_Exception Once this object was already stringified and closed.
+	 */
+	public function add_class( $class_name ) {
+		$this->assert_not_closed();
+		if ( null !== $this->tag_name_starts_at ) {
+			$this->classname_updates[ $class_name ] = self::ADD_CLASS;
+		}
+	}
+
+	/**
+	 * Removes a class name from the currently matched tag.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string $class_name The class name to remove.
+	 *
+	 * @throws WP_HTML_Walker_Exception Once this object was already stringified and closed.
+	 */
+	public function remove_class( $class_name ) {
+		$this->assert_not_closed();
+		if ( null !== $this->tag_name_starts_at ) {
+			$this->classname_updates[ $class_name ] = self::REMOVE_CLASS;
+		}
+	}
+
+	/**
+	 * Returns the string representation of the HTML Walker.
+	 * It closes the HTML Walker and prevents further lookups and modifications.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return string The processed HTML.
+	 */
+	public function __toString() {
+		if ( ! $this->is_closed() ) {
+			$this->after_tag();
+			$this->updated_html .= substr( $this->html, $this->updated_bytes );
+			$this->parsed_bytes  = strlen( $this->html );
+			$this->closed        = true;
+		}
+
+		return $this->updated_html;
+	}
```

Comment:
Hi! As an observer of this PR, I was curious about this implementation of `__toString()` — in particular, that it closes the walker. In my limited experience, I think it might be unexpected to developers that `__toString()` does something other than generate a string, and I was curious for the reasoning behind it. (I apologize if I've missed where it was described already.)

As an example of the unexpected behavior that I mean: In PHP 8+, this class will automatically be interpreted as implementing `Stringable`, such that it can be used with the `string|Stringable` type. See https://www.php.net/manual/en/class.stringable.php. 

However, using an instance of this class as a value where `string|Stringable` is the expected type will have a different effect than a plain string would once the value is used as a string. The plain string will still just be a string, but this object will start behaving differently, if I understand the code correctly.

Are there real-world cases where such confusion would arise? Not with WordPress core, of course, until it requires PHP 8. 

Until then, perhaps I want to debug the state of a walker object with a PSR-3 logger. See the type hints used in https://github.com/php-fig/log/blob/fe5ea303b0887d5caefd3d431c3e61ad47037001/src/LoggerInterface.php. 

If my object is logged while my transformations are in progress, wouldn't it stop working as expected after the logging occurred and the object was used as a string? The effect would be particularly tricky to diagnose if my object is logged only some of the time, such as if I log different error levels on production or lower environments .

As you can see, my starting position on this implementation isn't exactly neutral 😄 But I also recognize that this class was implemented to serve a specific purpose. Core also isn't necessarily obligated to consider potential integrations of core code with other libraries. In any case, I'm interested to learn!

Reply:
@dlh01 Spot on! We were discussing **not** closing the walker here recently specifically to avoid that confusion. Let's keep this discussion open until that happens.

Reply:
I just tried not closing the Walker and ran into problems.

I started with:

```php
	public function __toString() {
		return $this->updated_html . substr( $this->html, $this->updated_bytes );
	}
```

And quickly realized it doesn't apply the changes to the current tag – that happens in `$this->after_tag()`.

So I tried

```php
	public function __toString() {
		$this->after_tag();
		return $this->updated_html . substr( $this->html, $this->updated_bytes );
	}
```

Which work well but also finished processing the current tag, which means this code wouldn't work:

```php
$w->next_tag();
$w->set_attribute('id', 'foo');
$results = (string) $w;

// This line doesn't do anything:
$w->set_attribute('id', 'bar');
// Why? $this->current_tag has been processed and is now null.
```

The Walker only moves in a single direction. Applying a changes to a current tag means it cannot be updated anymore – that's a technical limitation.

The API could be explicit about that:

```php
$w->next_tag();
$w->set_attribute('id', 'foo');

// $results does **not** contain the id="foo" 
$results = (string) $w;

$w->next_tag(); // or $w->close_tag();

// $results **does** contain the id="foo" 
$results = (string) $w;
```

But it changes one gotcha for another – I'd expect to see the changes in `$result`. I find this quite confusing.

Perhaps the walker doesn't have to implement the `__toString()` method at all, then? For example:

```php
$w->next_tag();
$w->set_attribute('id', 'foo');

// $results **does** contain the id="foo" 
$results = $w->close_tag_and_apply_updates();
```

Or perhaps the walker could just expose the updated html so far for the sake of logging? For example:

```php
$w->next_tag();
$w->set_attribute('id', 'foo');

// Logs HTML without id="foo"
$logger->debug( $w->get_updated_html() );

// Closes and stringifies the walker
$results = (string) $w;
```

Reply:
Thanks for your response, @adamziel!

If there's a need for a method that closes the walker and returns the final string at the same time, then have that method. If `close_tag_and_apply_updates()` is cumbersome, then how about something like `final()` or `finalize()` or `close()`? I would only suggest: Don't use `__toString()`.

Logging is just an example use case. My focus isn't on logging specifically but on the expectations that a developer can have about a class that implements `__toString()`. My contention is that it's reasonable for a developer to expect that `__toString()` operates without side effects, just as plain strings don't have side effects.

Looking at one of the examples in your comment:

> Which worked well but it also finished processing the current tag, introducing the following issue:

```php
$w->next_tag();
$w->set_attribute('id', 'foo');
$results = (string) $w;

// This line doesn't do anything anymore:
$w->set_attribute('id', 'bar');
// Why? $this->current_tag has been processed and is now null.
```

Maybe I'm misreading the current code, but in the current implementation, won't the second `set_attribute()` not do anything either because the object has already been cast to a string, thus closing it?

Or, consider someone trying to do some dead-simple debugging:

```php
$w->set_attribute('id', 'foo');
echo "<pre>DEBUG: walker is currently {$w}</pre>";
$w->set_attribute('id', 'bar');
```

This will "work" because the class implements `__toString()`, but it will work differently once the debug statement is removed (again, if I'm reading correctly), and I think that will be unexpected to someone who sees that the class implements `__toString()`.

> Maybe the walker could just expose the updated html so far

IMO, if there's going to be a `__toString()` on the class, this is what it probably would return. If it's not safe, feasible, reasonable, etc. to expose the HTML prior to the class being closed, then, as you suggested, maybe there shouldn't be a `__toString()` at all.

(Another top-of-my-head option, although still somewhat awkward, would be for `__toString()` to return the HTML only after the class is explicitly closed via something like a dedicated `close()` method, and an empty string before then.)

Reply:
@adamziel when we discussed not closing the walker my point was we can stay where we are and for exactly the kinds of cases like this, particularly debugging each step. I think in those examples we're too focused on how it works at this moment and not on how we could do this without closing. of course it won't set another attribute after calling `__toString()` _if we close out the current tag_ but there's no need to close the current tag.

Right now we reset the current tag modifications after creating the string output but I think we could defer doing that until `next_tag()` and then if we continued to `set_attribute()` after calling `__toString()` then we would indeed do more processing than necessary, but could keep the interface the way it appears.

@dlh01 note that plenty of instances of WordPress will be running PHP 8+ long before it's required. Also I don't think this behavior is that different before or after PHP 8. Although we've chosen to add the explicit `(string)` cast in all these examples it's entirely unnecessary because when WordPress prints the walker, PHP calls `__toString()` automatically to perform the cast.

Reply:
@dmsnell ha, good idea! I removed the "closing" in [9bcfd3a](https://github.com/WordPress/gutenberg/pull/42485/commits/9bcfd3aeba7a9284113b0950886914439198c8c3) – `__toString()` now works quite intuitively:

```php
$w = new WP_HTML_Walker( '<div></div>' );
$w->next_tag();

$w->set_attribute('id', 'foo');
$results = (string) $w;
// <div id="foo"></div>

$w->set_attribute('id', 'bar');
$results = (string) $w;
// <div id="bar"></div>
```

The cost is doing additional work on each `__toString()`, which is fine. The developer has to decide if it's worth it.

Reply:
> note that plenty of instances of WordPress will be running PHP 8+ long before it's required. 

Of course. But WordPress core will not be able to use `Stringable` until it requires PHP 8, and I wanted to acknowledge that core developers, in the strictest interpretation, need not consider use cases beyond those in core.

> Also I don't think this behavior is that different before or after PHP 8. Although we've chosen to add the explicit (string) cast in all these examples it's entirely unnecessary because when WordPress prints the walker, PHP calls __toString() automatically to perform the cast.

Right, the internal behavior is the same after PHP 8, but I think the implications of a class defining `__toString()` change, as discussed above re. creating an expectation that `__toString()` behaves the same as a plain string.

I appreciate everyone's responses. With the latest updates, it does indeed look like `__toString()` behaves much more intuitively. Thank you all for the discussion!